File: src/main/java/ini/cx3d/spatialOrganization/SimpleTriangulationNodeOrganizer.java
/*
Copyright (C) 2009 Frédéric Zubler, Rodney J. Douglas,
Dennis Göhlsdorf, Toby Weston, Andreas Hauri, Roman Bauer,
Sabina Pfister & Adrian M. Whatley.

This file is part of CX3D.

CX3D is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CX3D is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CX3D.  If not, see <http://www.gnu.org/licenses/>.
*/

package ini.cx3d.spatialOrganization;

import ini.cx3d.spatialOrganization.interfaces.*;
import ini.cx3d.spatialOrganization.interfaces.SpaceNode;
import ini.cx3d.swig.simulation.SimpleTriangulationNodeOrganizerT_PhysicalNode;

import java.util.AbstractSequentialList;
import java.util.Iterator;
import java.util.LinkedList;

/**
 * This class is a very simple implementation of {@link AbstractTriangulationNodeOrganizer}.
 * All nodes are stored in a binary tree in order to obtain a good performance when
 * checking whether a certain node is already added to this organizer. 
 * @author Dennis Goehlsdorf
 *
 * @param <T> The type of user objects associated with the nodes in the current triangulation.
 */
public class SimpleTriangulationNodeOrganizer<T>  extends SimpleTriangulationNodeOrganizerT_PhysicalNode {
//		AbstractTriangulationNodeOrganizer<T> {

	BinaryTreeElement<T> treeHead = BinaryTreeElement.generateTreeHead();
	
	public SimpleTriangulationNodeOrganizer() {
//		registerJavaObject(this);
	}

	public void addTriangleNodes(ini.cx3d.spatialOrganization.interfaces.Triangle3D triangle) {
		addNode(triangle.getNodes()[1]);
		addNode(triangle.getNodes()[2]);
		addNode(triangle.getNodes()[0]);
	}

	public AbstractSequentialList<ini.cx3d.spatialOrganization.interfaces.SpaceNode> getNodes(final ini.cx3d.spatialOrganization.interfaces.SpaceNode referencePoint) {
		AbstractSequentialList<ini.cx3d.spatialOrganization.interfaces.SpaceNode> list = new LinkedList<>();
		Iterator<SpaceNode> it = getNodeIterator(referencePoint);
		while (it.hasNext()) {
			list.add(it.next());
		}
		return list;
	}

	@Override
	public String toString() {
		return "[" + treeHead + "]";
	}
	@Override
	public void addNode(ini.cx3d.spatialOrganization.interfaces.SpaceNode node) {
		treeHead.insert(node);
	}
	       
	@Override
	public Iterator<ini.cx3d.spatialOrganization.interfaces.SpaceNode> getNodeIterator(
			ini.cx3d.spatialOrganization.interfaces.SpaceNode referencePoint) {
		return treeHead.iterator();
	}

	@Override
	public void removeNode(ini.cx3d.spatialOrganization.interfaces.SpaceNode node) {
		treeHead.remove(node, null);
	}
	@Override
	public ini.cx3d.spatialOrganization.interfaces.SpaceNode<T> getFirstNode() {
		return treeHead.bigger.content;
	}

}

File: src/spatial_organization/simple_triangulation_node_organizer.cc
#include "spatial_organization/simple_triangulation_node_organizer.h"

#include <sstream>

#include "string_util.h"
#include "physics/physical_node.h"
#include "spatial_organization/triangle_3d.h"
#include "spatial_organization/binary_tree_element.h"

namespace bdm {
namespace spatial_organization {

template<class T>
std::shared_ptr<SimpleTriangulationNodeOrganizer<T>> SimpleTriangulationNodeOrganizer<T>::create() {
  return std::shared_ptr < SimpleTriangulationNodeOrganizer < T >> (new SimpleTriangulationNodeOrganizer());
}

template<class T>
SimpleTriangulationNodeOrganizer<T>::SimpleTriangulationNodeOrganizer()
    : tree_head_ { BinaryTreeElement < T > ::generateTreeHead() } {
}

template<class T>
SimpleTriangulationNodeOrganizer<T>::~SimpleTriangulationNodeOrganizer() {
  delete tree_head_;
}

template<class T>
std::list<SpaceNode<T>*> SimpleTriangulationNodeOrganizer<T>::getNodes(SpaceNode<T>* reference_point) {
  return tree_head_->inOrderTraversal();
}

template<class T>
void SimpleTriangulationNodeOrganizer<T>::removeNode(SpaceNode<T>* node) {
  tree_head_->remove(node, nullptr);
}

template<class T>
void SimpleTriangulationNodeOrganizer<T>::addNode(SpaceNode<T>* node) {
  tree_head_->insert(node);
}

template<class T>
SpaceNode<T>* SimpleTriangulationNodeOrganizer<T>::getFirstNode() const {
  return tree_head_->bigger_->content_;
}

template<class T>
std::string SimpleTriangulationNodeOrganizer<T>::toString() const {
  std::stringstream str_stream;
  str_stream << "[";
  str_stream << StringUtil::toStr(tree_head_);
  str_stream << "]";
  return str_stream.str();
}

template<typename T>
bool SimpleTriangulationNodeOrganizer<T>::equalTo(const std::shared_ptr<SimpleTriangulationNodeOrganizer<T>>& other) {
  return this == other.get();
}

template class SimpleTriangulationNodeOrganizer<bdm::physics::PhysicalNode> ;

}  // namespace spatial_organization
}  // namespace bdm

